File: next-billing-date-drifts/python/fix_next_billing_date.py
"""Realign a Shopify subscription contract's nextBillingDate when it drifts
from the date the billing policy actually implies.

The stored nextBillingDate on a SubscriptionContract should always be the
origin date plus a whole number of intervals (for example every 30 days, or
every 1 month). A paused-then-resumed contract, a manually edited date, or a
missed billing cycle can leave nextBillingDate sitting on a date the policy
never produces. This walks each active contract, recomputes the date the
policy implies, and calls subscriptionBillingCycleScheduleEdit to move the
next cycle back onto schedule when it drifts past a small tolerance.
Read only apart from the one write. Run on a schedule. Safe to run again
and again.
"""
import os
import logging
from datetime import datetime, timezone, date

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("fix_next_billing_date")

SHOP = os.environ["SHOPIFY_SHOP"]
TOKEN = os.environ["SHOPIFY_ACCESS_TOKEN"]
API_VERSION = os.environ.get("SHOPIFY_API_VERSION", "2025-01")
ENDPOINT = f"https://{SHOP}/admin/api/{API_VERSION}/graphql.json"
DRIFT_TOLERANCE_DAYS = int(os.environ.get("DRIFT_TOLERANCE_DAYS", "1"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

ACTIVE_STATUSES = {"ACTIVE"}

# Shopify's billingPolicy.interval values.
DAY_LENGTHS = {"DAY": 1, "WEEK": 7, "MONTH": 30, "YEAR": 365}

CONTRACTS_QUERY = """
query($cursor: String) {
  subscriptionContracts(first: 50, after: $cursor) {
    pageInfo { hasNextPage endCursor }
    nodes {
      id
      status
      nextBillingDate
      createdAt
      billingPolicy { interval intervalCount }
    }
  }
}"""

SET_NEXT_BILLING_DATE = """
mutation($contractId: ID!, $date: DateTime!) {
  subscriptionContractSetNextBillingDate(contractId: $contractId, date: $date) {
    contract { id nextBillingDate }
    userErrors { field message }
  }
}"""


def gql(query, variables=None):
    r = requests.post(
        ENDPOINT,
        json={"query": query, "variables": variables or {}},
        headers={"X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def _parse_date(value):
    """Accept a date or datetime string and return a date object."""
    text = value[:10]
    return date.fromisoformat(text)


def expected_next_billing_date(origin_date, interval, interval_count, today):
    """Walk forward from the origin in whole intervals and return the first
    cycle date that is on or after today. Pure: no I/O, no now() lookups.
    """
    step_days = DAY_LENGTHS.get(interval)
    if not step_days or interval_count <= 0:
        return None
    span = step_days * interval_count
    origin_ordinal = origin_date.toordinal()
    today_ordinal = today.toordinal()
    if today_ordinal <= origin_ordinal:
        return origin_date
    elapsed = today_ordinal - origin_ordinal
    cycles_passed = elapsed // span
    candidate_ordinal = origin_ordinal + cycles_passed * span
    if candidate_ordinal < today_ordinal:
        candidate_ordinal += span
    return date.fromordinal(candidate_ordinal)


def decide_realignment(contract, today):
    """Pure decision function. Given a contract dict and a reference today
    date, return the ISO date string to write, or None if nothing to do.
    """
    if contract.get("status") not in ACTIVE_STATUSES:
        return None

    policy = contract.get("billingPolicy") or {}
    interval = policy.get("interval")
    interval_count = policy.get("intervalCount")
    if not interval or not interval_count:
        return None

    stored_raw = contract.get("nextBillingDate")
    origin_raw = contract.get("createdAt")
    if not stored_raw or not origin_raw:
        return None

    stored = _parse_date(stored_raw)
    origin = _parse_date(origin_raw)

    expected = expected_next_billing_date(origin, interval, interval_count, today)
    if expected is None:
        return None

    drift_days = abs((stored - expected).days)
    if drift_days <= DRIFT_TOLERANCE_DAYS:
        return None

    return expected.isoformat()


def set_next_billing_date(contract_id, iso_date):
    result = gql(SET_NEXT_BILLING_DATE, {"contractId": contract_id, "date": iso_date})[
        "subscriptionContractSetNextBillingDate"
    ]
    if result["userErrors"]:
        raise RuntimeError(result["userErrors"])
    return result["contract"]["nextBillingDate"]


def active_contracts():
    cursor = None
    while True:
        data = gql(CONTRACTS_QUERY, {"cursor": cursor})["subscriptionContracts"]
        for node in data["nodes"]:
            yield node
        if not data["pageInfo"]["hasNextPage"]:
            return
        cursor = data["pageInfo"]["endCursor"]


def run():
    today = datetime.now(timezone.utc).date()
    fixed = 0
    for contract in active_contracts():
        target = decide_realignment(contract, today)
        if target is None:
            continue
        log.info(
            "Contract %s drifted. stored=%s expected=%s. %s",
            contract["id"], contract.get("nextBillingDate"), target,
            "would realign" if DRY_RUN else "realigning",
        )
        if not DRY_RUN:
            set_next_billing_date(contract["id"], target)
        fixed += 1
    log.info("Done. %d contract(s) %s.", fixed, "to realign" if DRY_RUN else "realigned")


if __name__ == "__main__":
    run()

File: next-billing-date-drifts/node/fix-next-billing-date.js
/**
 * Realign a Shopify subscription contract's nextBillingDate when it drifts
 * from the date the billing policy actually implies.
 *
 * The stored nextBillingDate on a SubscriptionContract should always be the
 * origin date plus a whole number of intervals (for example every 30 days,
 * or every 1 month). A paused-then-resumed contract, a manually edited date,
 * or a missed billing cycle can leave nextBillingDate sitting on a date the
 * policy never produces. This walks each active contract, recomputes the
 * date the policy implies, and calls subscriptionContractSetNextBillingDate
 * to move the next cycle back onto schedule when it drifts past a small
 * tolerance. Read only apart from the one write. Run on a schedule.
 * Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/shopify/next-billing-date-drifts/
 */
import { pathToFileURL } from "node:url";

const SHOP = process.env.SHOPIFY_SHOP || "example.myshopify.com";
const TOKEN = process.env.SHOPIFY_ACCESS_TOKEN || "shpat_dummy";
const API_VERSION = process.env.SHOPIFY_API_VERSION || "2025-01";
const ENDPOINT = `https://${SHOP}/admin/api/${API_VERSION}/graphql.json`;
const DRIFT_TOLERANCE_DAYS = Number(process.env.DRIFT_TOLERANCE_DAYS || 1);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const ACTIVE_STATUSES = new Set(["ACTIVE"]);

// Shopify's billingPolicy.interval values.
const DAY_LENGTHS = { DAY: 1, WEEK: 7, MONTH: 30, YEAR: 365 };
const MS_PER_DAY = 24 * 60 * 60 * 1000;

function parseDateOnly(value) {
  // Accept a date or datetime string, keep only the calendar date, in UTC.
  const text = value.slice(0, 10);
  const [y, m, d] = text.split("-").map(Number);
  return Date.UTC(y, m - 1, d);
}

function toIsoDate(ms) {
  return new Date(ms).toISOString().slice(0, 10);
}

/**
 * Walk forward from the origin in whole intervals and return the first
 * cycle date (as epoch ms, UTC midnight) that is on or after today.
 * Pure: no I/O, no Date.now() lookups.
 */
export function expectedNextBillingDateMs(originMs, interval, intervalCount, todayMs) {
  const stepDays = DAY_LENGTHS[interval];
  if (!stepDays || !intervalCount || intervalCount <= 0) return null;
  const spanMs = stepDays * intervalCount * MS_PER_DAY;
  if (todayMs <= originMs) return originMs;
  const elapsed = todayMs - originMs;
  const cyclesPassed = Math.floor(elapsed / spanMs);
  let candidateMs = originMs + cyclesPassed * spanMs;
  if (candidateMs < todayMs) candidateMs += spanMs;
  return candidateMs;
}

/**
 * Pure decision function. Given a contract object and a reference "today"
 * (epoch ms, UTC midnight), return the ISO date string to write, or null
 * if nothing needs to change.
 */
export function decideRealignment(contract, todayMs) {
  if (!ACTIVE_STATUSES.has(contract.status)) return null;

  const policy = contract.billingPolicy || {};
  const { interval, intervalCount } = policy;
  if (!interval || !intervalCount) return null;

  const storedRaw = contract.nextBillingDate;
  const originRaw = contract.createdAt;
  if (!storedRaw || !originRaw) return null;

  const storedMs = parseDateOnly(storedRaw);
  const originMs = parseDateOnly(originRaw);

  const expectedMs = expectedNextBillingDateMs(originMs, interval, intervalCount, todayMs);
  if (expectedMs === null) return null;

  const driftDays = Math.abs(storedMs - expectedMs) / MS_PER_DAY;
  if (driftDays <= DRIFT_TOLERANCE_DAYS) return null;

  return toIsoDate(expectedMs);
}

async function gql(query, variables = {}) {
  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers: { "X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Shopify ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const CONTRACTS_QUERY = `
query($cursor: String) {
  subscriptionContracts(first: 50, after: $cursor) {
    pageInfo { hasNextPage endCursor }
    nodes {
      id
      status
      nextBillingDate
      createdAt
      billingPolicy { interval intervalCount }
    }
  }
}`;

const SET_NEXT_BILLING_DATE = `
mutation($contractId: ID!, $date: DateTime!) {
  subscriptionContractSetNextBillingDate(contractId: $contractId, date: $date) {
    contract { id nextBillingDate }
    userErrors { field message }
  }
}`;

async function* activeContracts() {
  let cursor = null;
  while (true) {
    const data = (await gql(CONTRACTS_QUERY, { cursor })).subscriptionContracts;
    for (const node of data.nodes) yield node;
    if (!data.pageInfo.hasNextPage) return;
    cursor = data.pageInfo.endCursor;
  }
}

async function setNextBillingDate(contractId, isoDate) {
  const result = (
    await gql(SET_NEXT_BILLING_DATE, { contractId, date: isoDate })
  ).subscriptionContractSetNextBillingDate;
  if (result.userErrors.length) throw new Error(JSON.stringify(result.userErrors));
  return result.contract.nextBillingDate;
}

export async function run() {
  const todayMs = Date.UTC(
    new Date().getUTCFullYear(),
    new Date().getUTCMonth(),
    new Date().getUTCDate()
  );
  let fixed = 0;
  for await (const contract of activeContracts()) {
    const target = decideRealignment(contract, todayMs);
    if (target === null) continue;
    console.log(
      `Contract ${contract.id} drifted. stored=${contract.nextBillingDate} expected=${target}. ${
        DRY_RUN ? "would realign" : "realigning"
      }`
    );
    if (!DRY_RUN) await setNextBillingDate(contract.id, target);
    fixed++;
  }
  console.log(`Done. ${fixed} contract(s) ${DRY_RUN ? "to realign" : "realigned"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
